### PR TITLE
Add Equipment support

### LINKED ISSUE #662: Entity Equipment
## Describe the problem related to your feature request.

Valence does not sync player equipment (held items, armor, etc)

## What solution would you like?
a Equipment plugin which exposes a Equipment component.

## What alternative(s) have you considered?
this could also be implemented in the inventory crate, although in my opinion it would be better to implement equipment as a seperate Plugin.

## Additional context
related #254

I believe an Equipment Component could be implemented like this

```rs
#[derive(Debug, Default, Clone, Component)]
pub struct Equipment {
    main_hand: ItemStack,
    off_hand: ItemStack,
    boots: ItemStack,
    leggings: ItemStack,
    chestplate: ItemStack,
    helmet: ItemStack,

    /// Contains a set bit for each modified slot in `slots`.
    #[doc(hidden)]
    pub changed: u8,
}

impl Equipment {
    pub const SLOT_COUNT: u8 = 5;
    
    pub const MAIN_HAND_SLOT: u8 = 0;
    pub const OFF_HAND_SLOT: u8 = 1;
    pub const BOOTS_SLOT: u8 = 2;
    pub const LEGGINGS_SLOT: u8 = 3;
    pub const CHESTPLATE_SLOT: u8 = 4;
    pub const HELMET_SLOT: u8 = 5;
    
    pub fn main_hand(&self) -> &ItemStack {
        self.main_hand;
    }
    
    pub fn set_main_hand(&self) {
         ...
     }
     ...
}
```

that would make it pretty easy to use.

I also think this should be as seperate as possible from ``valence_inventory`` (so people using this crate would have to implement custom logic to update equipment based on inventory changes, but that should not be to hard)

we would have to use Inventory to get the HeldItem

thats the packet https://wiki.vg/index.php?title=Protocol&oldid=18375#Set_Equipment

Edit:
since we would have to use setters/getters (so we garuantee that updated the changed field)
we might as well store the actual items in a list instead of the fields.

Edit 2:
My idea was that valence would only provide this component, and an updater system.
So you would have to implement custom logic if you want to lets say set your equipment when you put on armor (as i do think there are use cases where you dont want to bind it to the inventory)

Any thoughts on that?

### PR DESCRIPTION
**Pull request by maxomatic458**
# Objective

- make valence work with entity equipment (armor, main/off hand items)
# Solution

adds the crate `valence_equipment` that exposes the Equipment Plugin.
every `LivingEntity` will have a `Equipment` Component.

The Equipment plugin will NOT be compatible with the inventory by default (thats intended, as that makes it possible to use the equipment feature without the inventory feature + I do think there are probably use cases where you want to make players appear as having armor, although they dont have it in their inventory)

This PR would add Events when entities are (un)loaded by a player (used for sending equipment once the player loads an entity). I do believe this might also be a useful feature outside of the equipment feature.

used #254 as reference + stole example idea

fixes #662 

Opening as a draft for now (might refactor the updating/event emitting system).
Let me know what you think of the Entity Load/Unload events.

### CODE REVIEW

**Comment by maxomatic458** [comment]
if the `equipment` feature is enabled (which it is by default), every `LivingEntity` will have the `Equipment` Component.

**Comment by maxomatic458** [comment]
Should be ready now (not sure why the CI is failing, the EquipmentPlugin is used in the DefaultPlugins)

As said above, this touches the client code and implements Custom events there, kind of a hack, but that seems to be the easiest way to do it without having to reimplement the "view" logic.

**Review by dyc3 — COMMENTED** [review]
I haven't reviewed this in depth yet, but I like how this is decoupled from inventories. Good tests too.

I think it would be nice to provide a utility system to keep equipment in sync with the entity's inventory. I foresee that being common enough that we should just provide that.

**Comment by maxomatic458** [comment]
> I think it would be nice to provide a utility system to keep equipment in sync with the entity's inventory. I foresee that being common enough that we should just provide that.

I agree with that for player inventories (as players can change the inventory themselves).

But im not sure that would be good for all `LivingEntities`, that would open up 2 ways of dealing with equipment (either through the `Equipment`, or through the `Inventory` component.
And i believe it would be better to just allow equipment handling through the `Equipment` component.

I can't think of any case where you would have to handle anything equipment related through the inventory

**Comment by maxomatic458** [comment]
Ive added the `EquipmentInventorySync` component that will have to be attached to entities, once attached the inventory & equipment components will be synced (equipment changes will be prioritized if both change in the same tick).

File: crates/valence_equipment/src/lib.rs
Comment (dyc3) [inline crates/valence_equipment/src/lib.rs:1]: This should include readme.md as the module level docs. I'm pretty sure that's how we have it set up for other packages.

File: crates/valence_equipment/README.md
Comment (dyc3) [inline crates/valence_equipment/README.md:1]: Feel free to elaborate on these docs with code examples, explanations of equipment slots, links to other docs, etc. Consider what would be useful as an end user.

File: crates/valence_equipment/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    mut commands: Commands,
+    mut entities: Query<Entity, (Added<LivingEntity>, Without<Equipment>)>,
+) {
+    for entity in &mut entities {
+        commands.entity(entity).insert(Equipment::default());
+    }
+}
+
+#[cfg(feature = "inventory")]
+mod inventory_sync {
```
Comment (dyc3) [inline crates/valence_equipment/src/lib.rs:252]: Move this module to a new file

File: crates/valence_equipment/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+fn on_entity_init(
+    mut commands: Commands,
+    mut entities: Query<Entity, (Added<LivingEntity>, Without<Equipment>)>,
+) {
+    for entity in &mut entities {
+        commands.entity(entity).insert(Equipment::default());
+    }
+}
+
+#[cfg(feature = "inventory")]
```
Comment (dyc3) [inline crates/valence_equipment/src/lib.rs:251]: I don't think it's necessary to hide this behind a feature.

File: crates/valence_equipment/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                inventory_sync::on_attach_inventory_sync,
+            ),
+        )
+        .add_systems(
+            Update,
+            (
+                update_equipment.before(FlushPacketsSet),
+                on_entity_load.before(FlushPacketsSet),
+                #[cfg(feature = "inventory")]
+                inventory_sync::equipment_inventory_sync,
```
Comment (dyc3) [inline crates/valence_equipment/src/lib.rs:31]: We don't need to put this behind a feature, the system will not run if nothing matches the query.

File: crates/valence_equipment/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        app.add_systems(
+            PreUpdate,
+            (
+                on_entity_init,
+                #[cfg(feature = "inventory")]
+                inventory_sync::on_attach_inventory_sync,
+            ),
+        )
+        .add_systems(
+            Update,
```
Comment (dyc3) [inline crates/valence_equipment/src/lib.rs:26]: IIRC, we've reserved the Update stage to be *just* for end user systems. This should probably be PostUpdate.

**Review by dyc3 — CHANGES_REQUESTED** [review]
Looks good, just minor things and docs.

File: crates/valence_equipment/src/lib.rs (hunk 4 above)
Comment (maxomatic458) [reply crates/valence_equipment/src/lib.rs]: i had to move inventory_sync::equipment_inventory_sync to ``PreUpdate``, because the tests were being flaky.

**Comment by dyc3** [comment]
Go ahead and hit resolve on the comments you've addressed in new commits so I know you got them.

**Comment by maxomatic458** [comment]
should be ready now!

**Comment by maxomatic458** [comment]
i just noticed that the `HeldItem` is not marked as changed. (which means the InventorySync will not work for Helditems)
https://github.com/valence-rs/valence/blob/f526a175bb7fee13d59c6987d2bf4e9b532866e5/crates/valence_inventory/src/lib.rs#L1411


is there any reason for not updating it?
@AviiNL (i saw you made that change)

**Comment by maxomatic458** [comment]
yea looks like removing that causes some funky behavior.

Ill try to change that so we dont rely on the HeldItem, but listen for the `UpdateSelectedSlotEvent` instead

**Comment by AviiNL** [comment]
> i just noticed that the `HeldItem` is not marked as changed. (which means the InventorySync will not work for Helditems)
> 
> https://github.com/valence-rs/valence/blob/f526a175bb7fee13d59c6987d2bf4e9b532866e5/crates/valence_inventory/src/lib.rs#L1411
> 
> is there any reason for not updating it? @AviiNL (i saw you made that change)

I don't even remember what I had for breakfast this morning, let alone something I've done over a year ago lol, there probably was a reason for it since it was discussed and the implementation changed in PR #525 there might be more info in discord but I really can't remember.
This is also from pre 1.21, so it's very possible that 1.21 changes made this requirement (whatever it was) obsolete.

**Comment by lukashermansson** [comment]
> This is also from pre 1.21, so it's very possible that 1.21 changes made this requirement (whatever it was) obsolete.

We are still on 1.20, the work to get us to 1.21 is still ongoing, so as far as I'm aware any requirement that was present then should still be a requirement as of right now (in `main` at least)

**Comment by maxomatic458** [comment]
> yea looks like removing that causes some funky behavior.
> Ill try to change that so we dont rely on the HeldItem, but listen for the UpdateSelectedSlotEvent instead

I added a seperate system specifically for handling client HeldItem changes now

File: crates/valence_equipment/src/inventory_sync.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
+use valence_inventory::player_inventory::PlayerInventory;
+use valence_inventory::{HeldItem, Inventory, UpdateSelectedSlotEvent};
+use valence_server::entity::player::PlayerEntity;
+
+use super::*;
+#[derive(Debug, Default, Clone, Component)]
+pub struct EquipmentInventorySync;
```
Comment (dyc3) [inline crates/valence_equipment/src/inventory_sync.rs:7]: This is user facing and needs a doc comment.

**Review by dyc3 — APPROVED** [review]
Looks good, just a minor doc thing.

**Pull request merged**